crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -43,146 +41,292 @@ pub enum BalanceVariant {
     /// Default format: no symbol, no denomination
     Default(Balance),
     /// Denominated format: symbol and denomination are present
-    Denominated(String),
+    Denominated(DenominatedBalance),
 }
 
 #[derive(Debug, Clone)]
 pub struct TokenMetadata {
     /// Number of denomination used for denomination
-    pub denomination: u128,
+    pub denomination: usize,
     /// Token symbol
     pub symbol: String,
 }
 
+#[derive(Debug, Clone)]
+pub struct DenominatedBalance {
+    value: Decimal,
+    unit: UnitPrefix,
+    symbol: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum UnitPrefix {
+    Giga,
+    Mega,
+    Kilo,
+    One,
+    Milli,
+    Micro,
+    Nano,
+}
+
 impl TokenMetadata {
     /// Query [TokenMetadata] through the node's RPC
     pub async fn query(client: &Client) -> Result<Self> {
         let sys_props = client.rpc().system_properties().await?;
 
         let default_decimals = json!(12);
         let default_units = json!("UNIT");
-        let decimals = sys_props
+        let denomination = sys_props
             .get("tokenDecimal")
             .unwrap_or(&default_decimals)
             .as_u64()
-            .context("error converting decimal to u64")?;
+            .context("error converting decimal to u64")?
+            as usize;
         let symbol = sys_props
             .get("tokenSymbol")
             .unwrap_or(&default_units)
             .as_str()
             .context("error converting symbol to string")?;
-        let denomination: u128 = format!("1{}", "0".repeat(decimals as usize)).parse()?;
         Ok(Self {
             denomination,
             symbol: symbol.to_string(),
         })
     }
 }
 
+impl TryFrom<String> for DenominatedBalance {
```

Comment:
If you change this to  `FromStr` you can replace `parse_balance` with using plain `value_parser` (in clap `v4`)